lib/Data/GoogleCloudStorage.php

```
@@ -217,6 +209,88 @@ public function existsComment($pasteid, $parentid, $commentid)
+                        } catch (NotFoundException $e) {
+                            // deleted by another instance.
+                        }
+                    }
+                }
+            }
+        } catch (NotFoundException $e) {
+            // no objects in the bucket yet
+        }
+    }
+
+    /**
+     * For GoogleCloudStorage, the value will also be stored in the metadata for the
+     * namespaces traffic_limiter and purge_limiter.
+     * @inheritDoc
+     */
+    public function setValue($value, $namespace, $key = '')
+    {
+        if ($key === '') {
+            $key = 'config/' . $namespace;
+        } else {
+            $key = 'config/' . $namespace . '/' . $key;
+        }
+
+        $metadata = array('namespace' => $namespace);
+        if ($namespace != 'salt') {
+            $metadata['value'] = strval($value);
+        }
+        try {
+            $this->_bucket->upload($value, array(
+                'name'          => $key,
+                'chunkSize'     => 262144,
+                'predefinedAcl' => 'private',
+                'metadata'      => array(
+                    'content-type' => 'application/json',
+                    'metadata'     => $metadata,
+                ),
+            ));
+        } catch (Exception $e) {
+            error_log('failed to set key ' . $key . ' to ' . $this->_bucket->name() . ', ' .
+                trim(preg_replace('/\s\s+/', ' ', $e->getMessage())));
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getValue($namespace, $key = '')
+    {
+        if ($key === '') {
+            $key = 'config/' . $namespace;
+        } else {
+            $key = 'config/' . $namespace . '/' . $key;
+        }
+        try {
+            $o = $this->_bucket->object($key);
+            return $o->downloadAsString();
+        } catch (NotFoundException $e) {
+            return '';
```

Maybe log an error/warning at least?

Or is this error really expected?

ALso, if we ignore an exception, let's at least have a comment there that explains _why_ we do so.

---

AFAIK understood from the reference implementation (https://github.com/PrivateBin/PrivateBin/blob/1fd998f325d72ad0273c714ad39a441a41492730/lib/Data/Filesystem.php#L291), when the key is not found and empty string is returned.

The comment would be ` // no key was found` which seems a bit superfluous given the catch is for the NotFoundException.